go/enclave/rollupchain/rollup_chain.go

```diff
@@ -151,50 +147,22 @@ func (rc *RollupChain) ProduceGenesisRollup(blkHash common.L1RootHash) (*core.Ro
 }
 
 // ProcessL1Block is used to update the enclave with an additional L1 block.
-func (rc *RollupChain) ProcessL1Block(block types.Block, receipts types.Receipts, isLatest bool) (*common.BlockSubmissionResponse, *common.Header, error) {
+func (rc *RollupChain) ProcessL1Block(block types.Block, receipts types.Receipts, isLatest bool) (*common.L2RootHash, *core.Rollup, error) {
 	rc.blockProcessingMutex.Lock()
 	defer rc.blockProcessingMutex.Unlock()
 
 	// We update the L1 chain state.
 	err := rc.updateL1State(block, receipts, isLatest)
 	if err != nil {
-		return nil, nil, rc.rejectBlockErr(err)
+		return nil, nil, err
 	}
 
 	// We update the L1 and L2 chain heads.
-	newL2Head, blockStage, err := rc.updateL1AndL2Heads(&block)
+	newL2Head, producedRollup, err := rc.updateHeads(&block)
 	if err != nil {
-		return nil, nil, rc.rejectBlockErr(err)
-	}
-
-	// If we're the sequencer and we're beyond the genesis block, we produce and store the new L2 chain head.
```

Comment:
I pushed the production of the rollup down into `rc.updateHeads`, and managed to condense these various checks and conditions nicely.